tests/OptimizelyTest.php

```diff
@@ -1730,4 +1731,498 @@ public function testGetVariationBucketingIdAttribute()
+        //should return null and log a message when no feature flag found against a valid feature key
+        $this->loggerMock->expects($this->at(0))
+        ->method('log')
+        ->with(Logger::ERROR, "FeatureFlag Key \"{$feature_key}\" is not in datafile.");
+        $this->assertSame($this->optimizelyObject->isFeatureEnabled($feature_key, "user_id"), null);
+    }
+
+    public function testIsFeatureEnabledGivenInvalidFeatureFlag(){
+        // Create local config copy for this method to add error
+        $projectConfig = new ProjectConfig($this->datafile, $this->loggerMock, new NoOpErrorHandler());
+        $optimizelyObj = new Optimizely($this->datafile);
+
+        $config = new \ReflectionProperty(Optimizely::class, '_config');
+        $config->setAccessible(true);
+        $config->setValue($optimizelyObj, $projectConfig);
+
+        $feature_flag = $projectConfig->getFeatureFlagFromKey('mutex_group_feature');
+        // Add such an experiment to the list of experiment ids, that does not belong to the same mutex group
+        $experimentIds = $feature_flag->getExperimentIds();
+        $experimentIds [] = '122241';
+        $feature_flag->setExperimentIds($experimentIds);
+
+        //should return null when feature flag is invalid
+        $this->assertSame($optimizelyObj->isFeatureEnabled('mutex_group_feature', "user_id"), null);
+    }
+
+    public function testIsFeatureEnabledGivenFeatureFlagIsNotEnabledForUser(){
+        $optimizelyMock = $this->getMockBuilder(Optimizely::class)
+            ->setConstructorArgs(array($this->datafile, null, $this->loggerMock))
+            ->setMethods(array('sendImpressionEvent'))
+            ->getMock();
+
+        // should return false when no variation is returned for user
+        $decisionServiceMock = $this->getMockBuilder(DecisionService::class)
+            ->setConstructorArgs(array($this->loggerMock, $this->projectConfig))
+            ->setMethods(array('getVariationForFeature'))
+            ->getMock();
+
+        $decisionService = new \ReflectionProperty(Optimizely::class, '_decisionService');
+        $decisionService->setAccessible(true);
+        $decisionService->setValue($optimizelyMock, $decisionServiceMock);
+
+        $decisionServiceMock->expects($this->exactly(1))
+            ->method('getVariationForFeature')
+            ->will($this->returnValue(null));
+
+        $optimizelyMock->expects($this->never())
+            ->method('sendImpressionEvent');
+
+        $this->loggerMock->expects($this->at(0))
+            ->method('log')
+            ->with(Logger::INFO, "Feature Flag 'double_single_variable_feature' is not enabled for user 'user_id'.");
+
+        $this->assertSame(
+            $optimizelyMock->isFeatureEnabled('double_single_variable_feature', 'user_id'),
+             false);
+    }
+
+    public function testIsFeatureEnabledGivenFeatureFlagIsEnabledAndUserBeingExperimented(){
+        $optimizelyMock = $this->getMockBuilder(Optimizely::class)
```

Comment:
can we move $optimizelyMock & $decisionServiceMock to common place. because most of the time we set sendImpressionEvent mocked & getVariationForFeatureMocked.
I would suggest pass the method name you want ot mock & create a method for it and return the mocked instance. It's suggestion :)

Reply:
Simplicity comes at the cost of duplication